### ISSUE TITLE
avx512bw fails where sse41/avx2 succeed

### ISSUE THREAD

**Issue by keiranmraine** [bwa-mem2/bwa-mem2#53]
We have attempted to use `bwa-mem2` to process a large dataset.

*  ~18 Terabases of processing was successful
   *  549 "lanes"
   *  ~34 Gbases/lane
*  174 "lanes" failed with various memory errors or non-zero exit codes
*  Mapping to this version of [GRCh38](ftp://ftp.sanger.ac.uk/pub/cancer/dockstore/human/GRCh38_hla_decoy_ebv/core_ref_GRCh38_hla_decoy_ebv.tar.gz)
*  Execution was using the pre-compiled binaries from 2.0pre2 (on Ubuntu Bionic)
*  Mapping is successful under legacy `bwa mem`
*  All test runs were provided 120GB of memory on an otherwise empty host (didn't exceed 60GB)
* `./bin/bwa-mem2 mem -p ref/genome.fa data/mini.xi.gz > /dev/null`

Taking 1 of the cases the error messages change depending on number of CPUs:

* 1 thread : `double free or corruption (!prev)`
* 8 threads: `free(): invalid next size (normal)`
* 1 thread + valgrind = ***success*** - forced to AVX2

As it was working with AVX2, I ran tests on each of the binaries in isolation:

* sse41 - works
* avx2 - works
* avx512bw - fails

The host used supports avx512bw.

```
casm3-os0000001: grep avx512bw /proc/cpuinfo | uniq -c
     60 flags		: fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ss ht syscall nx pdpe1gb rdtscp lm constant_tsc rep_good nopl xtopology cpuid tsc_known_freq pni pclmulqdq ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand hypervisor lahf_lm abm 3dnowprefetch invpcid_single pti ssbd ibrs ibpb stibp fsgsbase tsc_adjust bmi1 hle avx2 smep bmi2 erms invpcid rtm mpx avx512f avx512dq rdseed adx smap clflushopt clwb avx512cd avx512bw avx512vl xsaveopt xsavec xgetbv1 arat pku ospke
```

With help from @mp15 we have a dataset that reliably reproduces this error with just 212 read pairs (151 length).

@mp15 has traced the problem with "Address Sanitiser" :

> Just a brief update as to where I have got to. It turns out that whilst Valgrind doesn't work on AVX-512 instructions, Address Sanitiser does and you can breakpoint the process at __asan_on_error@plt to see what is about to start writing over your stack/heap. I have managed to trace the issue we are seeing to the following line:
> 
> ```
> src/kswv.cpp:326
> mySeq1SoA[k * SIMD_WIDTH8 + j] = (seq1[k] == AMBIG_ ? AMBR:seq1[k]);
> ```
> 
> Which as far as I can tell is part of a AVX-512 banded smith-waterman implementation and I believe part of the mate pair rescue feature?
> 
> The array it is indexing into is a subunit of:
> 
> ```
> #define MAX_SEQ_LEN_REF_SAM 2048
> #define SIMD_WIDTH8 64
> 
> uint8_t *seq1SoA = NULL;
> seq1SoA = (uint8_t *)_mm_malloc(MAX_SEQ_LEN_REF_SAM * SIMD_WIDTH8 * numThreads * sizeof(uint8_t), 64);
> ```
> 
> via:
> 
> ```
> //#pragma omp parallel num_threads(numThreads)
>   {
>     int32_t i;
>     // uint16_t tid = omp_get_thread_num();
> uint16_t tid = 0;
>     uint8_t *mySeq1SoA = seq1SoA + tid * MAX_SEQ_LEN_REF_SAM * SIMD_WIDTH8;
> ```
> 
> Apparently the thread number here will always be 0 and I assume numThreads will always be 1? In anycase that makes the maximum array size: 2048*64*1=131072. The line in question is part of a loop:
> 
> ```
> for(k = 0; k < sp.len1; k++)
>         {
>           // mySeq1SoA[k * SIMD_WIDTH8 + j] = (seq1[k] == AMBIG_?0xFF:seq1[k]);
> mySeq1SoA[k * SIMD_WIDTH8 + j] = (seq1[k] == AMBIG_ ? AMBR:seq1[k]);
>         }
> ```
> 
> And when it errors, k is 2048 and j is 0, with sp.len1 equal to 20577 and past the end of the array, therefore it is likely to be sp.len1 at fault. I am now trying to figure out why this is happening but we may need to pass this to Intel and ask them to figure it out.

Unfortunately, I am unable to attach the data directly to the issue due to DAA requirements.
To apply for access please see here, I believe it requires you to be a member of an academic
institution though:

  https://www.sanger.ac.uk/legal/DAA/MasterController

Please let me know if there is any further information I can provide.

P.S.

The following have been observed on other datasets, I am still confirming if these are specific to avx512bw, so far all 174 are shown to work on sse41 and 81 on avx2 (incomplete on remainder):

```
bwa-mem2: malloc.c:2927: __libc_malloc: Assertion `!victim || chunk_is_mmapped (mem2chunk (victim)) || ar_ptr == arena_for_chunk (mem2chunk (victim))' failed.
bwa-mem2: malloc.c:3567: _int_malloc: Assertion `(fwd->size & NON_MAIN_ARENA) == 0' failed.
bwa-mem2: malloc.c:3722: _int_malloc: Assertion `(unsigned long) (size) >= (unsigned long) (nb)' failed.
```

**Comment by arun-sub** [comment]
I faced the same issue with AVX512 and included some test datasets and fixes in #51 . Could you please take a look and let me know if they work for you ?

**Comment by keiranmraine** [comment]
@arun-sub thanks, we started looking into this early last week and didn't see your new PR when I raised it.  Will look into testing this.

**Comment by keiranmraine** [comment]
@arun-sub unfortunately your patch alone did not solve the instance we were investigating. @mp15 will be raising a PR with an extension to your work which is successful.

**Comment by keiranmraine** [comment]
I have managed to create a dataset that triggers the same issue using reference sequence only:

[ref-mini.fq.gz](https://github.com/bwa-mem2/bwa-mem2/files/4255529/ref-mini.fq.gz)

**Comment by keiranmraine** [comment]
#51 will solve this issue.

**Comment by yuk12** [comment]
I have added the reallocs (similar to the one implemented by arun in pull request #51). let me know if you face any issues.

**Comment by keiranmraine** [comment]
Can I confirm, this is all on master?  I'm not missing visibility of a development branch or something?

**Comment by keiranmraine** [comment]
@yuk12 I can confirm that this now runs to completion on 8 data sets that previously failed.  It is significantly slower than the version @arun-sub provided from this branch:

https://github.com/arun-sub/bwa-mem2/commits/mem_corruption_fix

Specifically this commit:

https://github.com/arun-sub/bwa-mem2/tree/831d11c00127a36f83579e94a869c3b27c9db640

```
avx512bw_20200227/1:    CPU time :                                   50802.23 sec.
avx512bw_20200317/1:    CPU time :                                   114679.00 sec.
```

(equivalent hardware, 8 cpus)

**Comment by yuk12** [comment]
@keiranmraine Will it be possible for you to share a dataset to reproduce this? Thanks.

**Comment by keiranmraine** [comment]
@yuk12 I am unable to share this data.  We initially had this slow down with first fix provided by @arun-sub, aligning the allocations solved it:

https://github.com/bwa-mem2/bwa-mem2/pull/51/commits/831d11c00127a36f83579e94a869c3b27c9db640?w=1

**Comment by yuk12** [comment]
Cool. Thanks :)

**Comment by keiranmraine** [comment]
@yuk12 do you want another round of testing?

**Comment by yuk12** [comment]
Yes, please. Thanks in advance.

**Comment by keiranmraine** [comment]
@yuk12 results are in, much faster:

```
cgppipe@casm3-head1: grep 'CPU time' logs/avx512-20200317/*
logs/avx512-20200317/1:    CPU time :                                   114679.00 sec.
logs/avx512-20200317/2:    CPU time :                                   121971.41 sec.
logs/avx512-20200317/3:    CPU time :                                   134016.03 sec.
logs/avx512-20200317/4:    CPU time :                                   139938.64 sec.
logs/avx512-20200317/5:    CPU time :                                   133400.00 sec.
logs/avx512-20200317/6:    CPU time :                                   116532.14 sec.
logs/avx512-20200317/7:    CPU time :                                   139155.31 sec.

cgppipe@casm3-head1: grep 'CPU time' logs/avx512-20200330/*
logs/avx512-20200330/1:    CPU time :                                   101301.38 sec.
logs/avx512-20200330/2:    CPU time :                                   91917.00 sec.
logs/avx512-20200330/3:    CPU time :                                   119815.00 sec.
logs/avx512-20200330/4:    CPU time :                                   108887.12 sec.
logs/avx512-20200330/5:    CPU time :                                   107275.70 sec.
logs/avx512-20200330/6:    CPU time :                                   85909.55 sec.
logs/avx512-20200330/7:    CPU time :                                   121100.32 sec.
```

Dated folder based on date of commit used.

**Comment by yuk12** [comment]
@keiranmraine Thanks a ton.
Do you see any slowdown as compared to previous commits?

Before this run you mentioned following slowdown:
avx512bw_20200227/1:    CPU time :                                   50802.23 sec.
avx512bw_20200317/1:    CPU time :                                   114679.00 sec.
is this ~2x slower?  and now the new number is 101301.38 sec ?

**Comment by keiranmraine** [comment]
I forgot to compare back to that version.  But yes it does appear that this version is still significantly slower than the version provided by @arun-sub on Feb 27th.

https://github.com/bwa-mem2/bwa-mem2/issues/53#issuecomment-604511004

I don't know what else may be involved.

**Comment by arun-sub** [comment]
@keiranmraine, there was a bug in my realloc implementation in #51 which @yuk12 pointed out. This could lead to different results if your dataset performed reallocations. It would be better to use runtimes of an older commit of BWA-MEM2 as your reference.

**Comment by keiranmraine** [comment]
Okay, so the fastest run I have on avx512bw prior to that one is `118164 s.`  I'll run comparison of each implementation (sse41, avx2, avx512bw) of the current build to confirm.

**Comment by keiranmraine** [comment]
I've just found out why that is so slow, missed a replace and the recent stats are actually avx2.  Repeating all now.  Sorry.

**Comment by keiranmraine** [comment]
Corrected run time for 20203030 commits, this makes far more sense:

```
$ grep 'CPU time' logs/{sse41,avx2,avx512bw}-20200330/1
   logs/sse41-20200330/1:    CPU time :                     130462.00 sec.
    logs/avx2-20200330/1:    CPU time :                      94966.13 sec.
logs/avx512bw-20200330/1:    CPU time :                      74183.03 sec.
$ grep 'CPU time' logs/{sse41,avx2,avx512bw}-20200330/2
   logs/sse41-20200330/2:    CPU time :                     114574.68 sec.
    logs/avx2-20200330/2:    CPU time :                      86097.48 sec.
logs/avx512bw-20200330/2:    CPU time :                      81591.84 sec.
$ grep 'CPU time' logs/{sse41,avx2,avx512bw}-20200330/3
   logs/sse41-20200330/3:    CPU time :                     140110.00 sec.
    logs/avx2-20200330/3:    CPU time :                     134825.52 sec.
logs/avx512bw-20200330/3:    CPU time :                      91147.00 sec.
$ grep 'CPU time' logs/{sse41,avx2,avx512bw}-20200330/4
   logs/sse41-20200330/4:    CPU time :                     162803.00 sec.
    logs/avx2-20200330/4:    CPU time :                     118373.68 sec.
logs/avx512bw-20200330/4:    CPU time :                      91178.53 sec.
```

**Comment by yuk12** [comment]
Yes, these numbers make more sense. Thanks for the hard work @keiranmraine
